recipes-extended/xen/files/xl-shutdown-wait-for-domain-death.patch

```diff
@@ -0,0 +1,44 @@
+################################################################################
+SHORT DESCRIPTION:
+################################################################################
+Wait for domain death during shutdown.
+
+################################################################################
+LONG DESCRIPTION:
+################################################################################
+While waiting on shutdown, domain death (DOMAIN_DEATH) depicts complete domain
+annhilation whereas shutdown domain (SHUTDOWN_DOMAIN) depicts just the beginning
+of domain destruction. So, it is only required to wait on domain death which is
+more meaningful and correct.
+
+################################################################################
+CHANGELOG
+################################################################################
+Contributions by:
+Mahantesh Salimath <salimathm@ainfosec.com>
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+To be upstreamed
+
+################################################################################
+PATCHES
+################################################################################
+Index: xen-4.9.0/tools/xl/xl_vmcontrol.c
+===================================================================
+--- xen-4.9.0.orig/tools/xl/xl_vmcontrol.c
++++ xen-4.9.0/tools/xl/xl_vmcontrol.c
+@@ -232,12 +232,6 @@ static void wait_for_domain_deaths(libxl
+             libxl_evdisable_domain_death(ctx, deathws[event->for_user]);
+             count++;
+             break;
+-        case LIBXL_EVENT_TYPE_DOMAIN_SHUTDOWN:
```

Comment:
DOMAIN_DEATH and SHUTDOWN_DOMAIN are two separate events, and they're both relevant.   I'm inclined not to remove this code block because libxl is supposed to handle the SHUTDOWN_DOMAIN event according to the reason code Xen sent back to it.  Completely ignoring the event seems like a bad choice to me.

What is the impact of leaving this code block in?

Reply:
Actually this code path is triggered only when wait option (-w) is passed with 'xl shutdown', in which case it should wait for the domain to go down (DOMAIN_DEATH).  As DOMAIN_DEATH always comes after SHUTDOWN_DOMAIN, the wait function returns  without actually waiting for the domain to go down completely. Yes, as you said DOMAIN_DEATH and SHUTDOWN_DOMAIN are two separate events but the libxl code which is supposed to handle these events is unchanged, please see tools/xl/xl_vmcontrol.c:(line 1094 to line 1169). The code here is just to make the wait return only when the domain shutdown is complete with its death.